A few improvements, in particular, to be able to perform the setup after a fresh installation

ingokloecker opened the pull request.

ingokloecker committed:
Do not crash when building title before server (and company) have been set

ingokloecker committed:
Use formatted string literal (aka f-string) for title

ingokloecker committed:
Skip loading of bank accounts if setup is not completed

ingokloecker committed:
Store the last window location and restore it on the next start

This fixes an annoying problem with a multi-monitor setup where the
window was shown at the center of the combined screen, e.g. in case
of two screens with identical resolution the window was shown with left
half on right side of left monitor and with right half on the left side
of the right monitor. This is a known bug in the PySimpleGUI. This still
happens on the first start, but on subsequent starts the last remembered
window location is restored. Unfortunately, we cannot remember the
location just before the window is closed because window.read() returns
after the window is already closed.

Note: After reading the last location from the settings we have to
convert it explicitly to a tuple because Window() expects a tuple, but
UserSettings returns a list (because JSON only knows lists).

tillmo commented on menu.py:
```diff
@@ -571,23 +571,30 @@ def menus():
 
     # ------ Loop & Process button menu choices ------ #
     window.bring_to_front()
+    last_window_location = window.current_location(more_accurate=True)
```
here I get an error: current_location() got an unexpected keyword argument `more_accurate`. Which version of PySimpleGUI do you use?

ingokloecker replied:
The latest version I guess. I have set up erpnext-client from scratch on my laptop yesterday. `pip freeze` says
```
PySimpleGUI==4.47.0
```

tillmo replied:
aha, indeed, 4.47.0 works. What is the Pythonic way to ensure that users upgrade to this version?

ingokloecker replied:
I checked the release notes. This parameter was added in 4.47.0 and this version was released two days ago. I didn't realize this was brand-new.

About ensuring an upgrade, I think the version requirement should be added to the `requirements.txt`, and after every `git pull --rebase` `pip3 install -r requirements.txt` should be run. I'll add this.

tillmo commented on utils.py:
```diff
@@ -102,11 +102,9 @@ def get_file(title):
-
+    company = settings.get('-company-', 'Unbekannt')
+    server = settings.get('-server-', 'unbekannt')
+    return f'{TITLE}{company}@{server} {VERSION}'
```
what is the rationale behind this refactoring? Why 'Unbekannt'?

ingokloecker replied:
On a new installation, `settings['-company-']` evaluates to `None` and `str+None` gives an exception. Therefore, I'm using `.get()` with a `str` as default value. We could use something else.

The second change, i.e. using an f-string instead of multiple string concatenations improves readability and is the modern Python approach for string formatting. Adding strings with `+` has always been an anti-pattern in Python because it can be an incredible performance issue if many strings are added. The (old) pattern for adding many strings is `''.join([s1, s2, s3, ...])`. This is magnitudes faster than adding multiple strings because it happens completely in the fast C-implementation and building an array is also much much faster than adding strings. Basically, it's O(1) vs. O(n). For this small string the performance doesn't matter, but apart from better readability one should avoid the anti-pattern even for small numbers of strings, because other people will understandably simply copy the anti-pattern when they need to build a string from thousands of strings.

I guess with the usage of an f-string the first change could be reverted because an f-string will interpolate a `None` value to `'None'` instead of raising an exception.

tillmo replied:
OK, understand. Then I would prefer the 'Unbekannt' version.

ingokloecker committed:
Add wrapper for getting the current window location

Only use the "more accurate" location. Otherwise, return an invalid
location so that PySimpleGUI uses the default location.
The non-accurate location has a large vertical offset (height of window
title bar and menu bar) which makes it useless for storing and restoring
on next start.

ingokloecker committed:
Require PySimpleGUI 4.47.0

ingokloecker committed:
Try to update the Python environment when updating the application

If the update fails (e.g. because on Windows files cannot be overwritten),
then give the user manual instructions.
